src/app/map/leaflet-map/leaflet-map.component.ts

```diff
@@ -0,0 +1,239 @@
+import { Component, AfterViewInit, OnChanges, OnDestroy, Input, OnInit, HostListener, ApplicationRef, Injector, ComponentFactoryResolver } from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
+import { LeafletMapUtils } from './leaflet-map.utils';
+import { Project } from 'app/models/project';
+import { ProjectService } from 'app/services/project.service';
+import { ProjectPopupComponent } from 'app/map/leaflet-map/project-popup/project-popup.component';
+
+import 'leaflet';
+import 'leaflet.markercluster';
+import * as L from 'leaflet';
+
+declare module 'leaflet' {
+  export interface FeatureGroup<P = any> {
+    projectId: number;
+  }
+  export interface Marker<P = any> {
+    projectId: number;
+  }
+}
+
+const markerIconYellow = L.icon({
+  iconUrl: 'assets/images/marker-icon-yellow.svg',
+  // Retina Icon is not needed here considering we're using an SVG. Enable if you want to change to a raster asset.
+  // iconRetinaUrl: 'assets/images/marker-icon-2x-yellow.svg',
+  iconSize: [36, 36],
+  iconAnchor: [18, 36],
+  tooltipAnchor: [16, -28]
+});
+
+@Component({
+  selector: 'leaflet-map',
+  templateUrl: './leaflet-map.component.html',
+  styleUrls: ['./leaflet-map.component.scss']
+})
+
+export class LeafletMapComponent implements OnInit, AfterViewInit, OnChanges, OnDestroy {
+
+  @Input() project: Project;
+  @Input() zoom = 6;
+  @Input() thumbnail = false;
+
+  public loading = false;
+  readonly defaultBounds = L.latLngBounds([48, -139], [60, -114]); // all of BC
+  readonly maxBounds = L.latLngBounds([40, -150], [70, -110]); // all of BC
+  public projects: Array<Project> = [];
+  public selectedProject: Project = null;
+  private map: L.Map = null;
+  private markerList: Array<L.Marker> = [];
+  private markerClusterGroup = L.markerClusterGroup({
+    showCoverageOnHover: false,
+    maxClusterRadius: 40, // 0 disables
+  });
+
+  constructor(
+    private projectService: ProjectService,
+    private router: ActivatedRoute,
+    private appRef: ApplicationRef,
+    private injector: Injector,
+    private resolver: ComponentFactoryResolver
+  ) { }
+
+  ngOnInit() {
+    // Load the projects. If this is a thumbnail with a project set, ignore
+    if (!this.thumbnail) {
+
+      let routerProjId = this.router.snapshot.paramMap.get('project');
+
+      this.projectService.getAll().subscribe(results => {
+        if (results) {
```

Comment:
An empty array will still pass this check. Not sure if that is intended or not.

Reply:
An empty array is fine, if a little useless. Because nothing else important happens in the `if` it can be ignored in a 0 element array so I'll add a length check